aten/src/ATen/core/type.cpp

```diff
@@ -752,46 +752,90 @@ TupleTypePtr TupleType::createNamed(
       field_types, qualName, schema)); // NOLINT(modernize-make-shared)
 }
 
-void flatten_union(TypePtr& type, std::vector<TypePtr>& res) {
+  bool NoneType::isSubtypeOfExt(const TypePtr& rhs, std::ostream *why_not) const {
+    if (rhs->kind() == OptionalType::Kind) {
+      return true;
+    }
+    return Type::isSubtypeOfExt(rhs, why_not);
+  }
+
+// Remove nested DynamicTypes during the instantiation of Union or
+// Optional. This populates `types` with all the types found during
+// flattening
+void flatten_dynamic_type(TypePtr& type, std::vector<TypePtr>& types) {
   if (type->kind() == UnionType::Kind) {
     for (auto inner_type : type->expect<UnionType>()->types()) {
-      flatten_union(inner_type, res);
+      flatten_dynamic_type(inner_type, types);
     }
   }
   else if (type->kind() == OptionalType::Kind) {
     auto inner_type = type->expect<OptionalType>()->getElementType();
-    if (inner_type->kind() == UnionType::Kind) {
-      flatten_union(inner_type, res);
+    if (inner_type->kind() == UnionType::Kind || inner_type->kind() == OptionalType::Kind) {
+      flatten_dynamic_type(inner_type, types);
     }
     else {
-      res.emplace_back(inner_type);
+      types.emplace_back(inner_type);
     }
-    res.emplace_back(NoneType::get());
+    types.emplace_back(NoneType::get());
   }
   else {
-    res.emplace_back(type);
+    types.emplace_back(type);
   }
 }
 
-UnionType::UnionType(std::vector<TypePtr> types) :
-  Type(TypeKind::UnionType) {
-  // Flatten Unions of Unions while populating `types_`
-  for (auto type : types) {
-    flatten_union(type, types_);
-  }
+// After calling `flatten_dynamic_type`, standardize the vector and
+// remove duplicates
+void sort_and_filter(std::vector<TypePtr>& types) {
   // We want the elements to be sorted so we can easily compare two UnionType
   // objects for equality in the future
-  std::sort(types_.begin(), types_.end(),
-  [](const TypePtr a, const TypePtr b) -> bool {
-    return a->kind() > b->kind();
+  std::sort(types.begin(), types.end(),
+            [](const TypePtr a, const TypePtr b) -> bool {
+              return a->str() > b->str();
```

Comment:
Curious why str comp is used instead

Reply:
Elias cleverly pointed out that my previous comparison [would not guarantee stable sorting](https://github.com/pytorch/pytorch/pull/53180#discussion_r610957012). I don't need the `TypePtr`s sorted in any special order; I just need to guarantee that a given set of `TypePtr`s will always be sorted in the same order. So, sort by their serialized representation.